File: app/src/main/java/test/cn/example/com/androidskill/model/Manager.java
package test.cn.example.com.androidskill.model;

import test.cn.example.com.androidskill.inter.CallBackInter;
import test.cn.example.com.util.LogUtil;

/**
 * Created by xgxg on 2017/6/13.
 */

public class Manager implements CallBackInter{
    public Manager(Employee employee){
        employee.work(this);
    }
    @Override
    public void callBack() {
        LogUtil.i("老板，我的工作做完了");
    }
}

File: app/src/main/java/test/cn/example/com/androidskill/kotlin/chapter3/Manager.kt
package test.cn.example.com.androidskill.kotlin.chapter3

import test.cn.example.com.util.LogUtil

/**
 * Created by xugan on 2019/7/11.
 */
class Manager :Drive,Write{
    override fun write() {
        LogUtil.i("写毛线")
    }

    override fun drive() {
        LogUtil.i("开车")
    }
}

class SeniorManager(val driver:Drive,val writer:Write):Drive by driver,Write by writer   // by driver  接口代理

class OlderDriver:Drive{
    override fun drive() {
        LogUtil.i("专业老司机")
    }
}

class PPTWritger:Write{
    override fun write() {
        LogUtil.i("专业十年写ppt")
    }

}

interface Drive{
    fun drive()
}

interface Write{
    fun write()
}